### PR TITLE
feat: support int64 <=> int32 auto conversion

### LINKED ISSUE #1382: ✨[Feature] Automatic conversion for int32<->int64 in fallback
As we know, if there are some operators that torch-tensorrt doesn't support, the model will be partitioned into tensorrt and torch subgraphs. TensorRT doesn't support int64 value and will truncate int64 to int32.

In some cases, the operators in the torch subgraph consume int64 value(like aten::index), and this value is produced from tensorrt subgraph(truncated into int32), this will cause an error. We need to track the data type conversion and automatic convert the data type back to the origianl type between torch and tensorrt.

Here is a typical case
```python
import torch
import torch.nn as nn
import torch_tensorrt


class Net(nn.Module):
  def __init__(self):
    super(Net, self).__init__()

  def forward(self, data, index):
    src = 1
    index = index.to(torch.int64)
    data = data * data
    data = data.scatter_(1,index,src)
    data = data + 1
    return data

data = torch.randn([5,5])
index = torch.randint(0,4,[2,2], dtype = torch.int32)

compile_spec = {
    "inputs": None,
    "device": {
        "device_type": torch_tensorrt.DeviceType.GPU,
        "gpu_id": 0,
        "allow_gpu_fallback": False,
        "disable_tf32": False
    },
    "truncate_long_and_double": True,
    "require_full_compilation": False,
    "torch_executed_ops": ["aten::scatter_", "aten::scatter"],
    "min_block_size": 1
}

net = Net()
model = torch.jit.trace(net, (data, index))

torch_type = torch.float32
min_shape = [5,5]
data2 = torch_tensorrt.Input(shape=min_shape, dtype=torch_type)
torch_type = torch.int32
index2 = torch_tensorrt.Input(shape=min_shape, dtype=torch_type)

inputs = [data2, index2]

compile_spec["inputs"] = inputs

with torch_tensorrt.logging.debug():
    trt_mod = torch_tensorrt.ts.compile(model, **compile_spec)
inputs = [data.cuda(), index.cuda()]
output = trt_mod(*inputs)
print(output)
```
subgraph log
```
INFO: [Torch-TensorRT - Debug Build] - Partitioned Graph: [Segment Block @0:
    Target: TensorRT

    Graph: graph(%index.1 : Tensor,
      %data.1 : Tensor):
  %2 : int = prim::Constant[value=4]() # test_int64.py:28:0
  %3 : bool = prim::Constant[value=0]() # test_int64.py:28:0
  %4 : NoneType = prim::Constant()
  %index : Tensor = aten::to(%index.1, %2, %3, %3, %4) # test_int64.py:28:0
  %data.3 : Tensor = aten::mul(%data.1, %data.1) # test_int64.py:29:0
  return (%index, %data.3)

Segment Block @1:
    Target: Torch

    Graph: graph(%data.3 : Tensor,
      %index : Tensor):
  %2 : int = prim::Constant[value=1]() # test_int64.py:30:0
  %0 : Tensor = aten::scatter(%data.3, %2, %index, %2) # test_int64.py:30:0
  return (%0)

Segment Block @2:
    Target: TensorRT

    Graph: graph(%1 : Tensor):
  %2 : Tensor = prim::Constant[value={1}]() # test_int64.py:31:0
  %3 : int = prim::Constant[value=1]() # test_int64.py:30:0
  %0 : Tensor = aten::add(%1, %2, %3) # test_int64.py:31:0
  return (%0)

]
```

### PR DESCRIPTION
**Pull request by bowang007**
Signed-off-by: Bo Wang <bowa@nvidia.com>

# Description
Support int64 <=> int32 type conversion.

Fixes #1382 

## Type of change
- New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: core/partitioning/shape_analysis.cpp [hunk 1]
```diff
@@ -142,16 +163,56 @@ void getSegmentsOutputByRunning(
...
+          torch::jit::Node* cast_node = getUpstreamCastNode(seg_block.raw_inputs()[i]);
+          std::unordered_map<torch::jit::Value*, torch::jit::Value*> value_map;
+          value_map.insert({cast_node->inputs()[0], seg_block.inputs()[i]});
+          auto env = [&](torch::jit::Value* v) { return util::getOrAddInputForValue(v, inner_g, value_map); };
+          auto new_cast_node = inner_g->prependNode(inner_g->createClone(cast_node, env));
+          seg_block.inputs()[i]->replaceAllUsesAfterNodeWith(new_cast_node, new_cast_node->outputs()[0]);
+        }
+      }
+    }
+    // TODO: This part might be necessary for some model, now checkint to verify
```
Comment (narendasan) [inline core/partitioning/shape_analysis.cpp:185]: @bowang007 should this be uncommented?

**Comment by narendasan** [comment]
@inocsin Can you verify these changes on key models?

**Comment by narendasan** [comment]
Seems like tests are failing for partitioning?

**Comment by inocsin** [comment]
> @inocsin Can you verify these changes on key models?

Sure, we have asked users to test this pr with their models

File: core/partitioning/shape_analysis.cpp (hunk 1 above)
Comment (bowang007) [reply core/partitioning/shape_analysis.cpp]: yes, optimized and refactored, this part is now included.

**Comment by gs-olive** [comment]
When compiling BART (https://huggingface.co/facebook/bart-base) using Torch TensorRT, this PR currently segfaults on my machine. Will add an additional comment with the line which is causing the issue

File: core/partitioning/shape_analysis.cpp [hunk 2]
```diff
@@ -57,6 +58,62 @@ std::unordered_map<const torch::jit::Value*, torch::jit::IValue> generateRandomI
...
 }
 
+torch::jit::Node* getUpstreamCastNode(torch::jit::Value* val) {
+  std::queue<torch::jit::Value*> q;
+  q.push(val);
+  std::unordered_set<torch::jit::Node*> visited;
+  while (!q.empty()) {
+    auto cur_val = q.front();
+    q.pop();
+    auto node = cur_val->node();
```
Comment (gs-olive) [inline core/partitioning/shape_analysis.cpp:68]: For BART (https://huggingface.co/facebook/bart-base), the segfault occurs here, when the `cur_val` pointer is accessed. When `getSegmentsOutputByRunning` calls `createCastNode` on line 221, the index passed in is a `seg_blocks.output()` index, but on line 83, `createCastNode` uses that index for `seg_block.raw_inputs()`. For the block in question, the number of inputs and outputs differs, so this may contribute to the segfault.

File: core/partitioning/shape_analysis.cpp [hunk 3]
```diff
@@ -57,6 +58,62 @@ std::unordered_map<const torch::jit::Value*, torch::jit::IValue> generateRandomI
...
+      for (auto input : node->inputs()) {
+        q.push(input);
+      }
+    }
+  }
+  return nullptr;
+}
+
+torch::jit::Node* createCastNode(SegmentedBlock& seg_block, size_t index, bool is_input) {
+  torch::jit::Node* cast_node = getUpstreamCastNode(seg_block.raw_inputs()[index]);
```
Comment (gs-olive) [inline core/partitioning/shape_analysis.cpp:83]: Add an if/else here to use `raw_inputs()` if `is_input = true` otherwise use `raw_outputs()`

File: core/partitioning/shape_analysis.cpp [hunk 4]
```diff
@@ -57,6 +58,62 @@ std::unordered_map<const torch::jit::Value*, torch::jit::IValue> generateRandomI
...
+  auto g = seg_block.g();
+  // if we can find upstream aten::to node, we use it's parameters for creating new cast node
+  if (cast_node) {
+    std::unordered_map<torch::jit::Value*, torch::jit::Value*> value_map;
+    value_map.insert({cast_node->inputs()[0], g->inputs()[index]});
+    if (!is_input) {
+      // if this value is output, we need to cast it to int32
+      auto const_val = g->insertConstant(3);
+      value_map.insert({cast_node->inputs()[1], const_val});
```
Comment (gs-olive) [inline core/partitioning/shape_analysis.cpp:92]: Throws an error when the upstream `aten::to` node does not have `dtype` as its second argument. For example, the schema `aten::to.prim_Device(Tensor(a) self, Device? device, int? dtype=None, bool non_blocking=False, bool copy=False) -> Tensor(b|a)` has `Device` as its second value, and this insertion causes it to be transformed to an invalid schema. We need to differentiate between schemas to ensure the `dtype` is placed in the right position. It seems that valid schemas for `aten::to` have `dtype` as either the second or third argument, or not at all. I believe there should be a check should be in `getUpstreamCastNode` to see if dtype is any of the arguments, and then a second check here to see if it is second or third argument in the schema.

The check here could be something like an if/else checking the debugName at the second index, as in:
```cpp
if (cast_node->inputs()[1]->node()->output()->type()->kind() == torch::jit::TypeKind::DeviceObjType) {
  value_map.insert({cast_node->inputs()[2], const_val});
} else {
  value_map.insert({cast_node->inputs()[1], const_val});
}
```

File: core/partitioning/shape_analysis.cpp [hunk 5]
```diff
@@ -57,6 +58,62 @@ std::unordered_map<const torch::jit::Value*, torch::jit::IValue> generateRandomI
...
 
+torch::jit::Node* getUpstreamCastNode(torch::jit::Value* val) {
+  std::queue<torch::jit::Value*> q;
+  q.push(val);
+  std::unordered_set<torch::jit::Node*> visited;
+  while (!q.empty()) {
+    auto cur_val = q.front();
+    q.pop();
+    auto node = cur_val->node();
+    if (node->kind().toQualString() == std::string("aten::to")) {
```
Comment (gs-olive) [inline core/partitioning/shape_analysis.cpp:69]: May need an additional check to ensure that the `aten::to` schema is valid for `dtype` insertion, as some of these schemas do not take an integer `dtype` at all, for example:

- `aten::to(Tensor(a) self, bool non_blocking=False, bool copy=False) -> Tensor(b|a)` 
- `aten::to(Tensor(a) self, Device device, ScalarType dtype, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)`
- `aten::to(Tensor(a) self, Tensor other, bool non_blocking=False, bool copy=False, MemoryFormat? memory_format=None) -> Tensor(a)`

A check could be something like an additional `&&` with

```cpp
(node->inputs()[1]->node()->output()->type()->kind() == torch::jit::TypeKind::IntType) ||
(node->inputs()[2]->node()->output()->type()->kind() == torch::jit::TypeKind::IntType)
```

File: core/partitioning/shape_analysis.cpp (hunk 2 above)
Comment (gs-olive) [inline core/partitioning/shape_analysis.cpp:68]: For BART (https://huggingface.co/facebook/bart-base), the segfault occurs here, on line 68, when the `cur_val` pointer is accessed. When `getSegmentsOutputByRunning` calls `createCastNode` on line 221, the index passed in is a `seg_blocks.output()` index, but on line 83, `createCastNode` uses that index for `seg_block.raw_inputs()`. For the block in question, the number of inputs and outputs differs, so this may contribute to the segfault.

**Review by gs-olive — CHANGES_REQUESTED** [review]
With the suggested edits, PR is functioning for BART model and successfully casts Long tensors to Int tensors. Suggestions are related to bugs arising from `input` vs `output` checking and usage of different `aten::to` schemas

**Comment by ncomly-nvidia** [comment]
Fixes #1346

File: core/partitioning/shape_analysis.cpp (hunk 5 above)
Comment (bowang007) [reply core/partitioning/shape_analysis.cpp:69]: Hey @gs-olive Any reproducer for this? 
What I'm not sure about is that for getUpstreamNode() function when we pass in a int32 value will the first `cast` node be the cast node that casts this value to int64? If that's the case, then we don't need this check. 
In other words, is it possible that the first `cast` node involving the passed value is to cast some other value? If the first `cast` node is not the cast node that casts to int64, will the second `cast` node be what we want?

Comment (gs-olive) [reply core/partitioning/shape_analysis.cpp:69]: Hi @bowang007 - yes, this did throw an error on my end, when compiling the [BART](https://huggingface.co/docs/transformers/model_doc/bart) model with the changes made in [this PR](https://github.com/pytorch/TensorRT/pull/1416), to fix an unrelated issue. The linked PR uses the second `aten::to` listed above, which does not take an integer dtype, so inserting the integer dtype there throws an error. Specifically, these lines in the PR cause the issue: https://github.com/pytorch/TensorRT/blob/9e7822aa842ee50316f1cac5d9d841ed04353c15/core/lowering/passes/device_casting.cpp#L28-L29

My thought was that we do need this check you have, but maybe it should be more stringent - something like:

```cpp
    if ((node->kind().toQualString() == std::string("aten::to")) &&
        ((node->inputs()[1]->node()->output()->type()->kind() == torch::jit::TypeKind::IntType) ||
        (node->inputs()[2]->node()->output()->type()->kind() == torch::jit::TypeKind::IntType))) {
```

File: core/partitioning/shape_analysis.cpp [hunk 6]
```diff
@@ -57,6 +58,59 @@ std::unordered_map<const torch::jit::Value*, torch::jit::IValue> generateRandomI
...
+        q.push(input);
+      }
+    }
+  }
+  return nullptr;
+}
+
+torch::jit::Node* createCastNode(SegmentedBlock& seg_block, size_t index, bool is_input) {
+  auto cast_raw_value = is_input ? seg_block.raw_inputs()[index] : seg_block.raw_outputs()[index];
+  auto cast_subgraph_value = is_input ? seg_block.outputs()[index] : seg_block.outputs()[index];
```
Comment (gs-olive) [inline core/partitioning/shape_analysis.cpp:84]: Please update this line to:

```cpp
auto cast_subgraph_value = is_input ? seg_block.inputs()[index] : seg_block.outputs()[index];
```

Currently, it is using the outputs regardless of the truth value of `is_input`. With this change, the PR (used along with PR #1416 is working for compilation + inference with the BART model)

File: core/partitioning/shape_analysis.cpp (hunk 5 above)
Comment (gs-olive) [reply core/partitioning/shape_analysis.cpp:69]: @bowang007 Please let me know what you think about the above

**Review by gs-olive — CHANGES_REQUESTED** [review]
One additional bugfix requested and one minor optional comment, and then the PR is successful when used for compilation + inference on BART.

File: core/partitioning/shape_analysis.cpp (hunk 5 above)
Comment (bowang007) [reply core/partitioning/shape_analysis.cpp]: @gs-olive I got your point now, let me update this part.

**Review by gs-olive — APPROVED** [review]

**Comment by gs-olive** [comment]
Looks good! Fully functional now on BART!
